src/gt4py/next/iterator/transforms/cse.py

```diff
@@ -90,25 +101,33 @@ def remove_subexprs(self, nodes: typing.Iterable[ir.Node]) -> None:
             for node in nodes:
                 subexpr_data = self.subexprs.pop(node, None)
                 if subexpr_data:
-                    node_ids, _ = zip(*subexpr_data)
+                    node_ids, _ = zip(*subexpr_data.subexprs)
                     node_ids_to_remove |= set(node_ids)
             for subexpr_data in self.subexprs.values():
-                for _, collected_child_node_ids in subexpr_data:
+                for _, collected_child_node_ids in subexpr_data.subexprs:
                     collected_child_node_ids -= node_ids_to_remove
 
     @classmethod
     def apply(cls, node: ir.Node) -> dict[ir.Node, list[tuple[int, set[int]]]]:
         state = cls.State()
         obj = cls()
-        obj.visit(node, state=state)
-        # return subexpression in pre-order of the tree, i.e. the nodes closer to the root come
-        # first, and skip the root node itself
-        return {k: v for k, v in reversed(state.subexprs.items()) if k is not node}
+        obj.visit(node, state=state, depth=-1)
+        # Return subexpression such that the nodes closer to the root come first and skip the root
+        # node itself.
+        subexprs_sorted: list[tuple[ir.Node, "CollectSubexpressions.SubexpressionData"]] = sorted(
+            state.subexprs.items(), key=lambda el: el[1].max_depth
+        )
+        return {k: v.subexprs for k, v in subexprs_sorted if k is not node}
+
+    def generic_visit(self, *args, **kwargs):
+        depth = kwargs.pop("depth")
+        return super().generic_visit(*args, depth=depth + 1, **kwargs)
 
     def visit(self, node: ir.Node, **kwargs) -> None:  # type: ignore[override]  # supertype accepts any node, but we want to be more specific here.
         if not isinstance(node, SymbolTableTrait) and not _is_collectable_expr(node):
             return super().visit(node, **kwargs)
 
+        depth = kwargs["depth"]
```

Comment:
why not `pop` to avoid the funny pattern in 143?

Reply:
(and move into the condition for `if_`)

Reply:
The depth is used also in line 187 so I can not move it into the `if_` part. I could also pop it which would make line 143 nicer, but then I would need to manually forward the depth in the other parts. I equally dislike both, pick one if you like to :-)